compiler/pkg/openapi_generator/generator_test.go

```diff
@@ -460,4 +466,192 @@ var _ = Describe("Generator", func() {
+			// modifying the `name` type from `string` to `int32`
+			tmpFile := fmt.Sprintf("%s/%s.yaml", tmpDir, "foos")
+			err = os.WriteFile(tmpFile, []byte(newFooCRD), 0665)
+			Expect(err).NotTo(HaveOccurred())
+
+			// should fail when change in annotation
+			fooContent, err := os.ReadFile(tmpFile)
+			Expect(err).To(BeNil())
+
+			var crd extensionsv1.CustomResourceDefinition
+			err = yaml.Unmarshal(fooContent, &crd)
+			Expect(err).To(BeNil())
+
+			f, err := os.OpenFile(tmpFile, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+			Expect(err).To(BeNil())
+
+			ann := crd.Annotations["nexus"]
+			nexusAnnotation := &pkg_generator.NexusAnnotation{}
+			err = json.Unmarshal([]byte(ann), &nexusAnnotation)
+			Expect(err).To(BeNil())
+
+			// modify singleton field to `true` which leads to DM incompatible with previous version
+			nexusAnnotation.IsSingleton = true
+			annotationInByte, err := yaml.Marshal(nexusAnnotation)
+			Expect(err).To(BeNil())
+
+			crd.Annotations["nexus"] = string(annotationInByte)
+			serialized, err := yaml.Marshal(crd)
+			Expect(err).To(BeNil())
+
+			_, err = f.Write(serialized)
+			Expect(err).To(BeNil())
+
+			// should be unsuccessful due to the incompatibility of the new and old CDs + forcing an upgrade=false
+			err = gen.UpdateYAMLs(tmpDir)
+			Expect(err).To(BeNil())
+
+			oldCRDDir, err := exampleFileTempTestDir("foos.yaml")
+			Expect(err).NotTo(HaveOccurred())
+
+			err = generator.CheckBackwardCompatibility(oldCRDDir, tmpDir, false)
+			cleanTempTestDir(oldCRDDir)
+			Expect(err.Error()).To(Equal("datamodel upgrade failed due to incompatible datamodel changes: \n " +
+				"detected changes in model stored in foos\n\n" +
+				"spec changes: \n/spec/versions/name=v1/schema/openAPIV3Schema/properties/spec/properties\n  " +
+				"- one field removed:\n    " +
+				"password:\n      " +
+				"type: string\n      " +
+				"format: string\n    \n  \n\n" +
+				"/spec/versions/name=v1/schema/openAPIV3Schema/properties/spec/properties/name/format\n  " +
+				"± value change\n    " +
+				"- string\n    " +
+				"+ int32\n  \n\n" +
+				"/spec/versions/name=v1/schema/openAPIV3Schema/properties/spec/properties/name/type\n  " +
+				"± value change\n    " +
+				"- string\n    " +
+				"+ integer\n  \n\n" +
+				"/spec/versions/name=v1/schema/openAPIV3Schema/properties/spec/required/0\n  " +
+				"± value change\n    " +
+				"- password\n    " +
+				"+ changePassword\n  \n\n" +
+				"nexus annotation changes: \n/is_singleton\n  ± value change\n    - false\n    + true\n  \n\n"))
+
+			// should fail when CRD is removed in the new list
+			oldCRDDir, err = exampleFileTempTestDir("zoos.yaml")
+			Expect(err).NotTo(HaveOccurred())
+
+			err = generator.CheckBackwardCompatibility(oldCRDDir, tmpDir, false)
+			Expect(err.Error()).To(Equal("datamodel upgrade failed due to incompatible datamodel changes: \n \"foos\" is deleted\n"))
+			cleanTempTestDir(oldCRDDir)
+
+			// shouldn't fail when no crds exists
```

Comment:
We probably need cases for multiple CRD's being present in the same file.
Both success / failure / one of them deleted cases ?

Reply:
@xmen4xp  Firstly, current compiler does not even support it. It generates a single CRD file each Nexus node.

Reply:
Let's say,

For node `Root `in pkg `root`  https://github.com/vmware-tanzu/graph-framework-for-microservices/blob/main/compiler/example/datamodel/root.go#L8

Compiler generates CRD file named <root_root.yaml> with pkg name and node name, in compatibility check, we are expecting the new CRD file with same name as existing one.

https://github.com/vmware-tanzu/graph-framework-for-microservices/blob/main/compiler/pkg/generator/template_renderers.go#L431

Reply:
@PavitraS makes sense.
So why do we have code to handle seperate multiple CRD's in a single file ?

Reply:
Clarified offline.